README.md
# Bank Simulation

Creating A bank account simulation. The following process was created:

- Create an account
- Get balance
- Deposit Money
- Withdraw Money
- Transfer Money
- Also added ability to use 3 currencies CAD, USD and EUROS with CAD been the primary currency

Unit test was added to test all functions created
Functional test added also

### Improvement scope

- Ability to dymically get current exchange rate for USD to CAD and EUROs to CAD


CHANGELOG.md
# Changelog

## [0.1.1](https://github.com/fixer-coder/banking-system/compare/v0.1.0...v0.1.1) (2023-12-28)


### Bug Fixes

* adding functional test and removing 3.7 python from unit test ([ade0c24](https://github.com/fixer-coder/banking-system/commit/ade0c241833291521762447cf2a78967d1adc76b))

## 0.1.0 (2023-12-28)

### Features

- Adding Bank account simulation ([5986ed3](https://github.com/fixer-coder/banking-system/commit/5986ed3af4ab3eea26226824f9bb64506eb0c702))

### Bug Fixes

- Adding linter to the repository ([f64dd77](https://github.com/fixer-coder/banking-system/commit/f64dd7744cd7fcb5eb43e6aef5006eb42188b8e0))
- Adding release please to the repository ([743c74f](https://github.com/fixer-coder/banking-system/commit/743c74f0eeffb696103e190e1ab7563602e0585e))


tests/functional_tests/test_bank.py
"""
This is a functional test of the scenarios A and B
"""
import logging

import pytest
from src.BankAccount import AceBank

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S %p",
)


@pytest.fixture
def setup_case_a():
    ab = AceBank(logging)
    ab.create_account(account_number="0808", balance=1000.00, customer_id=234)
    return ab


@pytest.fixture
def setup_case_b():
    ab = AceBank(logging)
    # setup account for customer 756
    ab.create_account(account_number="0903", balance=100.00, customer_id=756)
    ab.create_account(account_number="0875", balance=6000.00, customer_id=756)
    return ab


class TestCases(object):
    def test_case_a(self, setup_case_a):
        # deposit money for account 0808
        setup_case_a.deposit_fund("0808", 500.00, "USD", 234)

        # withdraw from account 0808
        setup_case_a.withdraw_funds("0808", 100.00, "CAD", 234)

        # get balance from account 0808
        balance = setup_case_a.get_balance("0808", 234)
        assert balance == 1650.00

    def test_case_b(self, setup_case_b):
        # withdraw from account 0875
        setup_case_b.withdraw_funds("0875", 700.00, "USD", 756)
        # deposit fund to 0903
        setup_case_b.deposit_fund("0903", 2500.00, "EUROS", 756)

        # transfer fund from 0875 to 0903
        setup_case_b.transfer_funds("0903", "0875", 1100.00, 756, 756)

        # get balance from account 0808
        balance_0875 = setup_case_b.get_balance("0875", 756)
        balance_0903 = setup_case_b.get_balance("0903", 756)
        assert balance_0875 == 3850.0
        assert balance_0903 == 6200.0


tests/unit_test/test_bank_account.py
"""
Unit test for AceBank class and all the functions in it
"""
import logging

import pytest
from src.BankAccount import AceBank

logging.basicConfig(
    filename="bank_app.logs",
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S %p",
)

ACCEPTED_CURRENCY = ["USD", "EUROS", "CAD"]


@pytest.fixture
def mock_bank():
    account_details = {
        250: {"8012": {"balance": 1000.00}, "7095": {"balance": 32500.95}},
        650: {"4000": {"balance": 350.10}},
    }

    ab = AceBank(logging, account_details=account_details)
    return ab


@pytest.fixture
def ace_bank():
    return AceBank(logging)


class TestAceBank(object):
    def test_create_account(self, ace_bank):
        """
        This test creates an account function
        """
        ace_bank.create_account("0912", "10000", 50)
        ace_bank.validate_account("0912", 50)
        assert ace_bank.account_details == {50: {"0912": {"balance": 10000}}}
        assert ace_bank.account_details[50] == {"0912": {"balance": 10000}}
        assert ace_bank.account_details[50]["0912"] == {"balance": 10000}

    def test_get_balance(self, ace_bank):
        # test no account
        with pytest.raises(ValueError) as no_account:
            ace_bank.get_balance("1010", 50)
        assert "Account does not exist" in str(no_account.value)

        # do proper testing to get balance
        ace_bank.create_account("08912", "10000.0", 50)
        acct_balance = ace_bank.get_balance("08912", 50)
        assert acct_balance == 10000.0

    def test_deposit_fund(self, mock_bank):
        # test deposit fund by using us dollars
        assert mock_bank.account_details[250]["8012"]["balance"] == 1000.00
        mock_bank.deposit_fund("8012", 200.00, "USD", 250)
        assert mock_bank.account_details[250]["8012"]["balance"] == 1300.00

        # test deposit fund using EUROS
        assert mock_bank.account_details[250]["8012"]["balance"] == 1300.00
        mock_bank.deposit_fund("8012", 200.00, "EUROS", 250)
        assert mock_bank.account_details[250]["8012"]["balance"] == 1700.00

        # test deposit fund using cad
        assert mock_bank.account_details[250]["8012"]["balance"] == 1700.00
        mock_bank.deposit_fund("8012", 200.00, "CAD", 250)
        assert mock_bank.account_details[250]["8012"]["balance"] == 1900.00

    def test_withdraw_funds(self, mock_bank):
        # test deposit fund using EUROS
        assert mock_bank.account_details[250]["8012"]["balance"] == 1000.00
        mock_bank.withdraw_funds("8012", 200.00, "EUROS", 250)
        assert mock_bank.account_details[250]["8012"]["balance"] == 600.00

        # test deposit fund by using us dollars
        assert mock_bank.account_details[650]["4000"]["balance"] == 350.10
        mock_bank.withdraw_funds("4000", 200.00, "USD", 650)
        assert mock_bank.account_details[650]["4000"]["balance"] == 50.10

        # test deposit fund by using us dollars
        assert mock_bank.account_details[250]["7095"]["balance"] == 32500.95
        mock_bank.withdraw_funds("7095", 2593.75, "CAD", 250)
        assert mock_bank.account_details[250]["7095"]["balance"] == 29907.20

    def test_transfer_fund(self, mock_bank):
        # confirm how much was in each account before transfer
        assert mock_bank.account_details[250]["7095"]["balance"] == 32500.95
        assert mock_bank.account_details[250]["8012"]["balance"] == 1000.00
        # make transfer
        mock_bank.transfer_funds("8012", "7095", 2000.00, 250, 250)
        # confirm changes
        assert mock_bank.account_details[250]["7095"]["balance"] == 30500.95
        assert mock_bank.account_details[250]["8012"]["balance"] == 3000.00

    def test_validate_account(self, mock_bank):
        # test account that does not exist
        with pytest.raises(ValueError) as va:
            mock_bank.validate_account("1010", 250)
        assert "Account does not exist" in str(va.value)

        # test account that exist
        mock_bank.validate_account("4000", 650)

    def test_validate_balance(self):
        # test when account_balance is greater than amount
        AceBank(logging).validate_balance(1000, 400)

        # test when account_balance is equal amount
        AceBank(logging).validate_balance(400, 400)

        # test when account_balance is less than amount
        with pytest.raises(ValueError) as low_balance:
            AceBank(logging).validate_balance(300, 400)
        assert "Account balance of 300 is currently lesser than 400" in (
            str(low_balance.value)
        )

    @pytest.mark.parametrize(
        "account_number, amount, currency, outgoing_acct_num",
        [(200.00, 200, 1, 854.2), (200, "error", "cad", 2)],
    )
    def test_validate_variable(
        self, account_number, amount, currency, outgoing_acct_num
    ):
        ac = AceBank(logging)
        # test when account_number
        with pytest.raises(ValueError) as wrong_variable:
            ac.validate_variable(account_number=account_number)
        assert f"Account number {account_number} must be string" in str(
            wrong_variable.value
        )

        # test when amount
        with pytest.raises(ValueError) as wrong_variable:
            ac.validate_variable(amount=amount)
        assert "Amount must be of type float" in str(wrong_variable.value)

        # test when currency
        with pytest.raises(ValueError) as wrong_variable:
            ac.validate_variable(currency=currency)
        assert (
            f"Currency {currency} is not recognized. Should be one of "
            f"{ACCEPTED_CURRENCY}" in str(wrong_variable.value)
        )

        # test when outgoing_account_number
        with pytest.raises(ValueError) as wrong_variable:
            ac.validate_variable(outgoing_account_number=outgoing_acct_num)
        assert f"Account number {outgoing_acct_num} must be string" in str(
            wrong_variable.value
        )

    def test_currency_converter(self):
        ac = AceBank(logging)
        # test converting us dollars to cad
        assert ac.currency_converter(100, "USD") == 150

        # test converting EUROS to cad
        assert ac.currency_converter(100, "EUROS") == 200


src/BankAccount.py
# Account simulation
import collections
import logging
import os
import sys

ACCEPTED_CURRENCY = ["USD", "EUROS", "CAD"]

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S %p",
)


class AceBank(object):
    def __init__(self, log=None, account_details=None):
        # keeping account info in a list
        self.account_details = account_details
        if self.account_details is None:
            self.account_details = collections.defaultdict(dict)
        self.log = log
        if self.log is None:
            self.log = logging.getLogger(os.path.basename(sys.argv[0]))

    def create_account(self, account_number, balance, customer_id):
        # This creates new customer account and add info to account details
        try:
            if self.validate_account(account_number, customer_id):
                raise Exception("Account does not exist")
        except ValueError:
            self.log.warning("We expect account should not exist so cont")
        self.account_details[customer_id][account_number] = {"balance": float(balance)}

    def get_balance(self, account_number, customer_id):
        self.validate_account(account_number, customer_id)
        current_acct = self.account_details[customer_id][account_number]
        return current_acct["balance"]

    def deposit_fund(self, account_number, amount, currency, customer_id):
        self.validate_variable(currency=currency)
        amount = self.currency_converter(amount, currency)
        self.validate_account(account_number, customer_id)
        self.account_details[customer_id][account_number]["balance"] = (
            self.account_details[customer_id][account_number]["balance"] + amount
        )

    def withdraw_funds(self, account_number, amount, currency, customer_id):
        self.validate_variable(
            account_number=account_number, amount=amount, currency=currency
        )
        self.validate_account(account_number, customer_id)
        amount = self.currency_converter(amount, currency)
        self.account_details[customer_id][account_number]["balance"] = round(
            self.account_details[customer_id][account_number]["balance"] - amount, 2
        )

    def transfer_funds(
        self,
        receivers_account_number,
        senders_account_number,
        amount,
        receivers_customer_id,
        senders_customer_id,
    ):
        self.validate_variable(
            account_number=receivers_account_number,
            amount=amount,
            outgoing_account_number=senders_account_number,
        )
        self.validate_account(senders_account_number, senders_customer_id)
        self.validate_account(receivers_account_number, receivers_customer_id)
        self.validate_balance(
            self.account_details[senders_customer_id][senders_account_number][
                "balance"
            ],
            amount,
        )

        # remove amount from sender's account
        self.account_details[senders_customer_id][senders_account_number]["balance"] = (
            self.account_details[senders_customer_id][senders_account_number]["balance"]
            - amount
        )

        # send amount to new account
        self.account_details[receivers_customer_id][receivers_account_number][
            "balance"
        ] = (
            self.account_details[receivers_customer_id][receivers_account_number][
                "balance"
            ]
            + amount
        )

    def validate_account(self, account_number, customer_id):
        # this validates that an account exist or not
        # if account exist returns account information
        # if account does not exist return false
        try:
            if self.account_details[customer_id][account_number]:
                self.log.info("Account exists")
        except KeyError:
            raise ValueError("Account does not exist")

    @staticmethod
    def validate_balance(account_balance, amount):
        if account_balance < amount:
            raise ValueError(
                f"Account balance of {account_balance} is currently lesser than {amount}"
            )

    @staticmethod
    def validate_variable(
        account_number=None, amount=None, currency=None, outgoing_account_number=None
    ):
        if account_number and not isinstance(account_number, str):
            raise ValueError(f"Account number {account_number} must be string")
        if amount and not isinstance(amount, float):
            raise ValueError("Amount must be of type float")
        if currency and currency not in ACCEPTED_CURRENCY:
            raise ValueError(
                f"Currency {currency} is not recognized. Should be one of {ACCEPTED_CURRENCY}"
            )
        if outgoing_account_number and not isinstance(outgoing_account_number, str):
            raise ValueError(f"Account number {outgoing_account_number} must be string")

    @staticmethod
    def currency_converter(amount, currency):
        if currency == "USD":
            amount = float(amount * 1.50)

        if currency == "EUROS":
            amount = float(amount * 2.00)
        return amount


script.py
# Testing AceBank Class in src/AceBank
import logging
import os
import sys

from src.BankAccount import AceBank

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S %p",
)
LOG = logging.getLogger(os.path.basename(sys.argv[0]))

if __name__ == "__main__":
    # initiate AceBank with customer id 234
    ab = AceBank(LOG)
    # setup account for customer 234
    ab.create_account(account_number="0808", balance=1000.00, customer_id=234)
    # deposit money for account 0808
    ab.deposit_fund("0808", 500.00, "USD", 234)

    # withdraw from account 0808
    ab.withdraw_funds("0808", 100.00, "CAD", 234)

    # get balance from account 0808
    balance = ab.get_balance("0808", 234)
    LOG.info(f"Balance of account 0808 after last transaction is {balance}")

    # initiate AceBank with customer id 756
    ab = AceBank(LOG)
    # setup accounts for customer 756
    ab.create_account(account_number="0903", balance=100.00, customer_id=756)
    ab.create_account(account_number="0875", balance=6000.00, customer_id=756)

    # withdraw from account 0875
    ab.withdraw_funds("0875", 700.00, "USD", 756)
    # deposit fund to 0903
    ab.deposit_fund("0903", 2500.00, "EUROS", 756)

    # transfer fund from 0875 to 0903
    ab.transfer_funds("0875", "0903", 1100.00, 756, 756)

    # get balance from account 0808
    balance_0875 = ab.get_balance("0875", 756)
    balance_0903 = ab.get_balance("0903", 756)
    LOG.info(f"Balance of account 0875 after last transaction is {balance_0875}")
    LOG.info(f"Balance of account 0903 after last transaction is {balance_0903}")
